### ISSUE TITLE
Consistent hash partitioning

### ISSUE THREAD

**Issue by prawn-cake** [postgrespro/pg_pathman#32]
Hi,

I'm curious about this extension so much. Currently it supports hash partitioning out of the box which is great. But for real world cases it's super-handy to have `partition key --> virtual bucket --> physical storage` kind of things to reduce re-sharding routine. 
I know about different approaches to work it around including smart clients or some manual table functions and mappings, but it seems to me the place where it could be done in same manner as existing partitioning ways with less client-side or manual routing table kind of things efforts, i.e "consistent hash partitioning" out of the box.

What's your take on this?

Thanks

**Comment by akorotkov** [comment]
Hi!

I know that virtual buckets are good for sharding.  This concept is used in citusdb and so on.  Besides pg_pathman can be used for FDW-based sharding, we're currently focusing on partitioning.
Probably, I didn't understood you correctly and this idea can give a benefit even for partition without distributing data between multiple database instances.  In this case, could you illustrate your idea in examples?
Anyway, we'll keep this issue in the mind during future development.

Thanks.

**Comment by prawn-cake** [comment]
@akorotkov you understood correctly, I kept sharding case in mind describing that. 
But it actually does not make much difference because the issue still remains the same. The issue is total data reorganization is required if more partitions need to be added. 
The issue can be formulated as change partitions_count dynamically after its creation with less efforts.

Let's say we decided to split a really big table with pg_pathman into 4 partitions with some hashable key, done it. But our data flow started growing increasingly and just in a while we came up with that we need add 4 more partitions (8 in total). We can't just add 4 more partitions, we break our "routing" to the old entries at least. We have to compose the data together and split it again into 8 partitions.
Having virtual buckets here we have to replace only small part (or smaller) to the new partitions instead of reorganizing full dataset.

Well, to be honest it doesn't look like feasible feature for the extension and rather looks like some mature database internals + proper routing of course. But it might give you some ideas in future.
